### DISCUSSION TITLE
prisma:generate produces absolute path

### DISCUSSION THREAD

**Discussion by hanar3** [prisma/prisma#15297] [Q&A] [answered]
Hi, sorry for the vague title! To summarize my use-case, I'm using prisma with two databases, so I have two schemas. In order to use prisma with two different databases, I have to generate the files and manually import them from the generated output.

So, whenever I update something for one of the schemas, I have to go:
`prisma generate --schema=./prisma/schema1.prisma`

And my schema would look like this:
```
generator client {
  provider      = "prisma-client-js"
  output        = "../src/generated/db-1"
  binaryTargets = ["native", "linux-musl"]
}
```

The problem:

My initial idea was to check the generated files into source control, since it's something that I only have to generate once in a while, it wouldn't be a problem, and others wouldn't have to go through the hassle of generating this themselves.

After prisma:generate, if you go to the generated folder and open `index.js`
```json
const config = {
  "generator": {
    "name": "client",
    "provider": {
      "fromEnvVar": null,
      "value": "prisma-client-js"
    },
    "output": {
      "value": "/Users/hanar3/Documents/testproject/src/generated/db1-client",
      "fromEnvVar": null
    },
   ....
}

```

See how the path on `output` is absolute? This causes prisma to throw on other environments, obviously due to the inexistent path. Question is - would it be possible for this to reference a relative path rather than an absolute one? Is this a config on the schema?

I can obviously just call prisma:generate again for other environments,  but this seems like it could be an unnecessary effort since it's only failing due to the absolute path...

There are many other paths which are relative on that same file, so I was wondering why this one can't be relative as well. Perhaps  the best practice here to re-generate everything?
Also, sorry if this is duplicate, I haven't found anything similar here...!

**Comment by austincrim** [comment, marked as answer]
Hey @hanar3 👋,

We generally recommend leaving generated files **out** of source control and instead regenerating the artifacts that can be regenerated. It typically results in a more robust experience.

**Reply by austincrim** [reply]
@hanar3 Would you mind also creating a [bug report](https://github.com/prisma/prisma/issues/new?assignees=&labels=kind%2Fbug&template=bug_report.yml) for this situation?

**Reply by hanar3** [reply]
Will do! To my point - Being able to "share" the generated code for my use-case would help a ton, and it feels to me that it could work if it was not for the way that path (and some others) are set up on the generated code. I've seen other generator-clis produce code, that _can_ be put into source control, namely openapi-generator-cli.

**Reply by jkhmtx** [reply]
Hi there, is there a continuation of this issue somewhere? I am having the same issue.

**Reply by pat-trunk-io** [reply]
This makes it impossible to have proper reproducible builds, the cache entry is different for every of our devs

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!

**Comment by vanenshi** [comment]
Hello Prisma
This error is preventing us from deploying the application using SST. Here is the information:

1. I generate Prisma using Github Actions. This is what the output looks like:
<img width="444" alt="image" src="https://github.com/user-attachments/assets/9c693973-7249-47fa-a458-283c06323e09" />

And here is my schema.prisam looks like 
```prisma
generator client {
  provider      = "prisma-client"
  output        = "./generated"
  binaryTargets = ["rhel-openssl-3.0.x", "native"]
}
```

And here is the error
```sh
Runtime.UnhandledPromiseRejection: PrismaClientInitializationError: Prisma Client could not locate the Query Engine for runtime "rhel-openssl-3.0.x".
            This is likely caused by tooling that has not copied "libquery_engine-rhel-openssl-3.0.x.so.node" to the deployment folder.
            Ensure that you ran `prisma generate` and that "libquery_engine-rhel-openssl-3.0.x.so.node" has been copied to "../../../../home/runner/work/thawapp/thawapp/packages/database/prisma/generated".
            We would appreciate if you could take the time to share some information with us.
            Please help us by answering a few questions: https://pris.ly/engine-not-found-tooling-investigation
            The following locations have been searched:
              /home/runner/work/thawapp/thawapp/packages/database/prisma/generated
              /var/task/node_modules/@prisma/client
              /var/task/node_modules/.prisma/client
              /tmp/prisma-engines
                at process.<anonymous> (file:///var/runtime/index.mjs:1294:17)
                at process.emit (node:events:524:28)
                at emitUnhandledRejection (node:internal/process/promises:250:13)
                at throwUnhandledRejectionsMode (node:internal/process/promises:385:19)
                at processPromiseRejections (node:internal/process/promises:470:17)
                at process.processTicksAndRejections (node:internal/process/task_queues:96:32)
```

Since SST uses the Open-Next to publish the Next.js application, absolute URLs here are causing a problem.

---
Right now, I am using this hack to publish the app, I use `../../../node_modules/.prisma/client` as output in the schema.Prisma file. And since Prisma falls back to look for the client files in there, it managed to find the generated files

**Comment by cyonkee** [comment]
Now with the move to typescript-based (rather than rust-based) client generation the previous general recommendation of "leaving generated files out of source control" does not seem applicable. We can now generate the clients in our source code and include the generated clients in our builds in a way that we never have to regenerate them unless the schema changes. This is possible because we don't have the limitation now of having to run the generation for specific architectures with our deployments. 

Is it possible this could get another look from the team?

This negatively impacts our ability to avoid rebuilding libs and caching in CI due to the absolute file path always changing. It makes builds non-deterministic.

This is further detailed in the prisma blog: https://www.prisma.io/blog/why-prisma-orm-generates-code-into-node-modules-and-why-it-ll-change
